pkg/engine/task/task_pipe.go

```diff
@@ -0,0 +1,452 @@
+	mountPath := ctr.VolumeMounts[0].MountPath
+
+	pod := corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: corev1.PodSpec{
+			Volumes: []corev1.Volume{
+				{
+					Name:         volumeName,
+					VolumeSource: corev1.VolumeSource{EmptyDir: nil},
+				},
+			},
+			InitContainers: []corev1.Container{*ctr},
+			Containers: []corev1.Container{
+				{
+					Name:    pipePodContainerName,
+					Image:   "busybox",
+					Command: []string{"/bin/sh", "-c"},
+					Args:    []string{"sleep infinity"},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      volumeName,
+							MountPath: mountPath,
+						},
+					},
+				},
+			},
+			RestartPolicy: corev1.RestartPolicyOnFailure,
+		},
+	}
+
+	b, err := yaml.Marshal(pod)
+	if err != nil {
+		return "", fmt.Errorf("failed to create pipe task pod: %v", err)
+	}
+
+	return string(b), nil
+}
+
+func copyFiles(fs afero.Fs, ff []PipeFile, pod *corev1.Pod, ctx Context) error {
+	restCfg, err := config.GetConfig()
+	if err != nil {
+		return fatalExecutionError(fmt.Errorf("failed to fetch cluster REST config: %v", err), pipeTaskError, ctx.Meta)
+	}
+
+	var g errgroup.Group
+
+	for _, f := range ff {
+		f := f
+		log.Printf("PipeTask: %s/%s copying pipe file %s", ctx.Meta.InstanceNamespace, ctx.Meta.InstanceName, f.File)
+		g.Go(func() error {
+			return copyFile(fs, f, pod, restCfg)
+		})
+	}
+
+	err = g.Wait()
+	return err
+}
+
+func copyFile(fs afero.Fs, pf PipeFile, pod *corev1.Pod, restCfg *rest.Config) error {
+	reader, stdout := io.Pipe()
+	stderr := bytes.Buffer{}
+
+	pe := PodExec{
+		RestCfg:       restCfg,
+		PodName:       pod.Name,
+		PodNamespace:  pod.Namespace,
+		ContainerName: pipePodContainerName,
+		Args:          []string{"tar", "cf", "-", pf.File},
+		In:            nil,
+		Out:           stdout,
+		Err:           &stderr,
+	}
+
+	// When downloading files using remotecommand.Executor, the call HAS to be wrapped into
+	// a goroutine, otherwise it blocks the execution. It boils down to calling remotecommand/v4.go:54
+	// method, which will try copy stdout and stderr into provided buffers. We're using io.Pipe to move data
+	// from remote stdout to the local reader. It works synchronously as it copies data directly from the
+	// Write to the corresponding Read, there is no internal buffering.
+	//
+	// IMPORTANT: the PodExec.Run call doesn't return until we consume data from the pipe reader! It has
+	// 			  to be executed in a goroutine and we have to consume the passed stdout first.
+	//            Otherwise this code will deadlock. This is why consuming error from the errCh has to
+	//            happen AFTER we try to extract the file! 🤯
+	//
+	// See `kubectl cp` copyFromPod method for another example:
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp/cp.go#L291
+	errCh := make(chan error, 0)
+	go func() {
+		defer stdout.Close()
```

Comment:
Please add a comment the `Close` for a `io.Pipe` never returns an error, hence we can ignore it here.